### PR TITLE
Refactor and Clean Up Theta Star Path Planner

### LINKED ISSUE #925: Refactor and Cleanup ThetaStarPathPlanner
### Description of the task
Currently, the ThetaStarPathPlanner has several issues:

- excessively long functions
- deep nesting (>2 levels) in a single function
- tons of arbitrary scaling factors
- very large single expressions spread out over 7+ lines
- very low test coverage (not just in terms of line coverage, particularly in terms of covering functionality, edge cases, etc. )

Related to #814

### Acceptance criteria

- [ ] Cleanup
- [ ] Item B
- [ ] Item C

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Clean up data structures, long functions, and coding style to make theta star path planner readable and ready for changes to make it better

### Testing Done
Regression tests

### Resolved Issues
Resolves #925 

### Length Justification
Mostly name changing

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 1]
```diff
@@ -34,133 +33,131 @@ bool ThetaStarPathPlanner::isUnBlocked(int row, int col)
...
 }
 
-double ThetaStarPathPlanner::calculateHValue(int row, int col, CellCoordinate dest)
+double ThetaStarPathPlanner::calculateHValue(Coordinate test_coord, Coordinate dest)
 {
     // Return using the distance formula
-    return ((double)sqrt((row - dest.first) * (row - dest.first) +
-                         (col - dest.second) * (col - dest.second)));
+    return (
+        (double)sqrt((test_coord.row() - dest.row()) * (test_coord.row() - dest.row()) +
```
Comment (EvanMorcom) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: Would this be better as a static_cast<double>?

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 2]
```diff
@@ -174,123 +171,145 @@ Path ThetaStarPathPlanner::findPath(const Point &start, const Point &destination
...
+        return Path(std::vector<Point>({start, closest_destination}));
     }
 
-    if ((start - destination).len() < CLOSE_TO_DEST_THRESHOLD ||
-        ((start - destination).len() < SIZE_OF_GRID_CELL_IN_METERS))
+    initListsAndCellDetails(src_coord);
+
+    bool found_dest = findPathToDestination(dest_coord);
+
+    if (found_dest == false)
```
Comment (EvanMorcom) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp:194]: This is only used once?

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 3]
```diff
@@ -453,44 +454,65 @@ Point ThetaStarPathPlanner::findClosestFreePoint(Point p)
...
-ThetaStarPathPlanner::CellCoordinate ThetaStarPathPlanner::convertPointToCell(Point p)
+ThetaStarPathPlanner::Coordinate ThetaStarPathPlanner::pointToCoordinate(Point p)
 {
     // account for robot radius
-    return std::make_pair((int)((p.x() + (fieldXHalfLength - ROBOT_MAX_RADIUS_METERS)) /
-                                SIZE_OF_GRID_CELL_IN_METERS),
-                          (int)((p.y() + (fieldYHalfLength - ROBOT_MAX_RADIUS_METERS)) /
-                                SIZE_OF_GRID_CELL_IN_METERS));
+    return Coordinate(
+        (int)((p.x() + max_navigable_x_coord) / SIZE_OF_GRID_CELL_IN_METERS),
```
Comment (EvanMorcom) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: Same casting comment as above

**Review by EvanMorcom — DISMISSED** [review]
💯

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h [hunk 4]
```diff
@@ -29,104 +29,123 @@ class ThetaStarPathPlanner : public PathPlanner
...
-     * @param row y position of cell
-     * @param col x position of cell
+     *
+     * @param test_coord Coordinate to consider
      * @param dest destination cell
      *
      * @return true if cell is the destination
      */
-    bool isDestination(int row, int col, CellCoordinate dest);
+    bool isDestination(Coordinate test_coord, Coordinate dest);
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.h]: The implementation of this function is just checking equality. Can we either change the function name to `isEqual` or just implement the equality operator for the `Coordinate` class? I was really confused at first why 2 arguments were needed to check if something was the destination.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

**Comment by jonathanlew** [comment]
Just some context for why things were so ugly in the first place is that I adapted it from https://www.geeksforgeeks.org/a-search-algorithm/ which was written in C and was not written cleanly

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h (hunk 4 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h]: can just use == here

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp:194]: I rather have the function call outside the if statement because findPathToDestination does some work

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: I don't think static cast is necessary

Comment (MathewMacDougall) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: According to our style guide we don't use c-style casts, we should always `static_cast` when possible. There's nitty-gritty reasons as to why it's safer.

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 5]
```diff
@@ -34,133 +33,132 @@ bool ThetaStarPathPlanner::isUnBlocked(int row, int col)
...
+                cell_heuristics[new_coord.row()][new_coord.col()].f      = f_new;
+                cell_heuristics[new_coord.row()][new_coord.col()].g      = g_new;
+                cell_heuristics[new_coord.row()][new_coord.col()].h      = h_new;
+                cell_heuristics[new_coord.row()][new_coord.col()].parent = parent_new;
             }
-            // If the destination GridCell is the same as the
+            // If the destination CellHeuristic is the same as the
             // current successor
-            if (isDestination(pNew.first, pNew.second, dest) == true)
+            if (isDestination(new_coord, dest) == true)
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: Were we going to replace `isDestination` with `==`?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 5 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: Yeah we can do this because Coordinates are pairs so == is well defined

Comment (MathewMacDougall) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: We're not using `==` here for the coordinates?

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: til

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 6]
```diff
@@ -174,123 +166,145 @@ Path ThetaStarPathPlanner::findPath(const Point &start, const Point &destination
...
+    max_navigable_y_coord = navigable_area.yLength() / 2.0 - ROBOT_MAX_RADIUS_METERS;
+    num_grid_rows = (int)((max_navigable_x_coord * 2.0 + ROBOT_MAX_RADIUS_METERS) /
+                          SIZE_OF_GRID_CELL_IN_METERS);
+    num_grid_cols = (int)((max_navigable_y_coord * 2.0 + ROBOT_MAX_RADIUS_METERS) /
+                          SIZE_OF_GRID_CELL_IN_METERS);
+
+    // Reset data structures to path plan again
+    open_list.clear();
+    closed_list.clear();
     unblocked_grid.clear();
```
Comment (garethellis0) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: Could we put all the member initialization in it's own function (so the "Initialize member variables" and "Reset data ..." sections here)?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
These changes are definitely helping a lot! 

I think we're still missing part of what was pointed out in #925, namely:
  - deep nesting: (`findClosestFreePoint` has 5 levels of indenting in at least one place, and there are lots of places we're 4 levels deep)
  - very low test coverage: doesn't look like we've added any more tests?

Lemme know if you want help with the above, I'm happy to make refactoring suggestions or figure out what's not really being tested right now.

Also, it seems like there are random tab characters in these files, could you please do a quick search and replace for all them? (And fix spacing after)

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: done

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h [hunk 7]
```diff
@@ -29,104 +29,113 @@ class ThetaStarPathPlanner : public PathPlanner
...
 
    private:
-    typedef std::pair<int, int> CellCoordinate;
-    typedef std::pair<double, CellCoordinate> OpenListCell;
+    class Coordinate : public std::pair<int, int>
+    {
+       public:
+        Coordinate(int row, int col) : std::pair<int, int>(row, col) {}
+
+        Coordinate() : std::pair<int, int>() {}
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: Coordinate extends pair so we can reuse all the operators that pair implements for data structures and comparators

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp [hunk 8]
```diff
@@ -34,133 +33,126 @@ bool ThetaStarPathPlanner::isUnBlocked(int row, int col)
...
             // the parent of this square. Record the
-            // f, g, and h costs of the square GridCell
+            // f, g, and h costs of the square CellHeuristic
             //			 OR
             // If it is on the open list already, check
             // to see if this path to that square is better,
             // using 'f' cost as the measure.
-            if (cellDetails[pNew.first][pNew.second].f_ == DBL_MAX ||
-                cellDetails[pNew.first][pNew.second].f_ > fNew)
+            if (cell_heuristics[new_coord.row()][new_coord.col()].f == DBL_MAX ||
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: I was considering changing this to a map. I will investigate that

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: Hmm, which ones specifically do we need?

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: To a map? Not quite sure that I follow. I'd also recommend against a map here, just because we do a _lot_ of lookups, and looking up a value at a known offset in a vector is much faster then doing the required hashing to lookup in a map.

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: All the operators we need for sets mainly

[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: Ok, so just hashing? Take a look at `point.h`, there's an example of how we made it work with unordered STL containers at the bottom.

[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: There are slightly different requirements for ordered containers, but should be trivial to implement the operators required I think?

[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: I see how point implements this. It seems like that's equivalent to how std::pair<int, int> would implement so I don't see how re-implementing all the operators, hash, etc. would be an improvement. What's your concern with extending std::pair<int, int>? Seems to me like a very concise way to represent a Coordinate.

**Comment by jonathanlew** [comment]
> These changes are definitely helping a lot!
> 
> I think we're still missing part of what was pointed out in #925, namely:
> 
>     * deep nesting: (`findClosestFreePoint` has 5 levels of indenting in at least one place, and there are lots of places we're 4 levels deep)

yeah... findClosestFreePoint is disgusting. We need rethink how that works because we would like merge overlapping obstacles via convex hulls and use that to find the closest free point, instead of just checking points in an expanding circle. With merged obstacles, we would be able to guarantee that a "blocked" point is inside one and only one obstacle. However, merging obstacles might be costly/hard to implement and would be part of new geom. Let's discuss that offline.

I'll look into the other 4 level nesting, as well.

> 
>     * very low test coverage: doesn't look like we've added any more tests?
> 

Hm... I think #770 will cover it so it shouldn't be part of #925. Any thetastar specific tests you'd suggest?
> 
> Lemme know if you want help with the above, I'm happy to make refactoring suggestions or figure out what's not really being tested right now.

I'll do another pass for more refactoring. As for tests, I think this PR would want to target tests that a generic path planner interface test wouldn't cover (because we have #770). Any suggestions for that?
> 
> Also, it seems like there are random tab characters in these files, could you please do a quick search and replace for all them? (And fix spacing after)

ok

**Comment by garethellis0** [comment]
Good point about the generic path planner test suite, I don't think there are any specific tests here (really we're validating that something obeys the path planner interface, and as long as Theta* does we should be fine).

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 6 above)
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: done

File: src/software/ai/navigator/path_planner/theta_star_path_planner.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.cpp]: done

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/navigator/path_planner/theta_star_path_planner.h:37]: Fair enough, I'm good with this

File: src/software/ai/navigator/path_planner/theta_star_path_planner.h [hunk 9]
```diff
@@ -22,111 +22,174 @@ class ThetaStarPathPlanner : public PathPlanner
...
+        CellHeuristic() {}
+
+        /**
+         * Updates and initializes CellHeuristics internal variables
+         *
+         * @param parent parent
+         * @param f f value
+         * @param g g value
+         */
+        void updateAndInitialize(Coordinate parent, double f, double g)
```
Comment (garethellis0) [inline src/software/ai/navigator/path_planner/theta_star_path_planner.h]: nit; `initialize` generally refers to the first time that you set a value, I would rename this to `update`.

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
:tada:
